sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -917,4 +846,127 @@ case class GpuCast(
+        withResource(checked.castTo(DType.create(DType.DTypeEnum.DECIMAL64, 0))) {
+          scaleZero => scaleZero.round(dt.scale, ai.rapids.cudf.RoundMode.HALF_UP)
+        }
+      } else if (dt.scale > 0) {
+        // Integer will be enlarged during casting if scale > 0, so we cast input to INT64
+        // before casting it to decimal in case of overflow.
+        withResource(checked.castTo(DType.INT64)) { long =>
+          long.castTo(DType.create(DType.DTypeEnum.DECIMAL64, -dt.scale))
+        }
+      } else {
+        checked.castTo(DType.create(DType.DTypeEnum.DECIMAL64, -dt.scale))
+      }
+    }
+  }
+
+  private def castFloatsToDecimal(input: ColumnVector, dt: DecimalType): ColumnVector = {
+
+    // Approach to minimize difference between CPUCast and GPUCast:
+    // step 1. cast input to FLOAT64 (if necessary)
+    // step 2. cast FLOAT64 to container DECIMAL (who keeps one more digit for rounding)
+    // step 3. perform HALF_UP rounding on container DECIMAL
+    val checkedInput = withResource(input.castTo(DType.FLOAT64)) { double =>
+      val roundedDouble = double.round(dt.scale, ai.rapids.cudf.RoundMode.HALF_UP)
+      withResource(roundedDouble) { rounded =>
+        // We rely on containerDecimal to perform preciser rounding. So, we have to take extra
+        // space cost of container into consideration when we run bound check.
+        val containerScaleBound = DType.DECIMAL64_MAX_PRECISION - (dt.scale + 1)
+        val bound = math.pow(10, (dt.precision - dt.scale) min containerScaleBound)
+        if (ansiMode) {
+          assertValuesInRange(rounded,
+            minValue = Scalar.fromDouble(-bound),
+            maxValue = Scalar.fromDouble(bound),
+            inclusiveMin = false,
+            inclusiveMax = false)
+          rounded.incRefCount()
+        } else {
+          replaceOutOfRangeValues(rounded,
+            minValue = Scalar.fromDouble(-bound),
+            maxValue = Scalar.fromDouble(bound),
+            inclusiveMin = false,
+            inclusiveMax = false,
+            replaceValue = Scalar.fromNull(DType.FLOAT64))
+        }
+      }
+    }
+
+    withResource(checkedInput) { checked =>
+      // If target scale reaches DECIMAL64_MAX_PRECISION, container DECIMAL can not
+      // be created because of precision overflow. In this case, we perform casting op directly.
+      if (DType.DECIMAL64_MAX_PRECISION == dt.scale) {
+        checked.castTo(DType.create(DType.DTypeEnum.DECIMAL64, -dt.scale))
+      } else {
+        val containerType = DType.create(DType.DTypeEnum.DECIMAL64, -(dt.scale + 1))
+        withResource(checked.castTo(containerType)) { container =>
+          container.round(dt.scale, ai.rapids.cudf.RoundMode.HALF_UP)
+        }
+      }
+    }
+  }
+
+  private def castDecimalToDecimal(input: ColumnVector,
+      from: DecimalType,
+      to: DecimalType): ColumnVector = {
+
+    val checkedInput = if (to.scale <= from.scale) {
```

Comment:
Can we add some comments here, because it is was not clear at all to me that we were doing the cast in two phases, where we first check the for overflows and then do the casting.